source/platforms/bitbucket_server/BitBucketServerAPI.ts

```diff
@@ -0,0 +1,196 @@
+import * as debug from "debug"
+import * as node_fetch from "node-fetch"
+import * as v from "voca"
+
+import {
+  BitBucketServerPRDSL,
+  BitBucketServerCommit,
+  BitBucketServerPRComment,
+  JIRAIssue,
+  BitBucketServerPRActivity,
+} from "../../dsl/BitBucketServerDSL"
+
+import { RepoMetaData } from "../../ci_source/ci_source"
+import { dangerSignaturePostfix, dangerIDToString } from "../../runner/templates/githubIssueTemplate"
+import { api as fetch } from "../../api/fetch"
+
+// Note that there are parts of this class which don't seem to be
+// used by Danger, they are exposed for Peril support.
+
+/** This represent the BitBucketServer API */
+
+export class BitBucketServerAPI {
+  fetch: typeof fetch
+  private readonly baseUrl = process.env["DANGER_BITBUCKETSERVER_HOST"]
+  private readonly d = debug("danger:BitBucketServerAPI")
+
+  private pr: BitBucketServerPRDSL
+
+  constructor(
+    public readonly repoMetadata: RepoMetaData,
+    public readonly repoCredentials: { host: string; username: string; password: string }
+  ) {
+    // This allows Peril to DI in a new Fetch function
+    // which can handle unique API edge-cases around integrations
+    this.fetch = fetch
+  }
+
+  private getPRBasePath(service = "api") {
+    const [projectKey, repositorySlug] = this.repoMetadata.repoSlug.split("/")
+    const pullRequestId = this.repoMetadata.pullRequestID
+    return (
+      `${this.baseUrl}/rest/${service}/1.0/` +
+      `projects/${projectKey}/` +
+      `repos/${repositorySlug}/` +
+      `pull-requests/${pullRequestId}`
+    )
+  }
+
+  getPullRequestInfo = async (): Promise<BitBucketServerPRDSL> => {
+    if (this.pr) {
+      return this.pr
+    }
+    const path = this.getPRBasePath()
+    const res = await this.get(path)
+    const prDSL = (await res.json()) as BitBucketServerPRDSL
+    this.pr = prDSL
+
+    if (res.ok) {
+      return prDSL
+    } else {
+      throw `Could not get PR Metadata for ${path}`
+    }
+  }
+
+  getPullRequestCommits = async (): Promise<BitBucketServerCommit[]> => {
+    const path = `${this.getPRBasePath()}/commits`
+    const res = await this.get(path)
+    return (await res.json()).values
+  }
+
+  getPullRequestDiff = async () => {
+    // TODO: possible?
```

Comment:
Can't seem to find an endpoint to get a raw `diff`, but there's a structured version in the REST API. Could use that but would require changing some common code.

Reply:
OK, I'm reasonably sure the two underlying features that rely on this are:

* Finding out the modified/created/deleted files
* Generating metadata for a specific file (e.g. maybe a JSON diff, or showing before after etc) 

There may be two different APIs that could get that kind of information from?